src/PathMappingRepository.php

```diff
@@ -0,0 +1,382 @@
+                $resource->attachTo($this, $path);
+
+                return $resource;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Search for resources by querying their path.
+     *
+     * @param string $query        The glob query.
+     * @param bool   $singleResult Should this method stop after finding a
+     *                             first result, for performances.
+     *
+     * @return ArrayResourceCollection The results of search.
+     */
+    private function search($query, $singleResult = false)
+    {
+        $resources = new ArrayResourceCollection();
+
+        /*
+         * If the query is not a glob, return it directly
+         */
+        if (!Glob::isDynamic($query)) {
+            $resource = $this->resolveResource($query);
+
+            if ($resource instanceof Resource) {
+                $resources->add($resource);
+            }
+
+            return $resources;
+        }
+
+        /*
+         * Otherwise, we need to search for paths matching the glob:
+         *      1.  find the root resource of all search results by
+         *          resolving the glob base path
+         *      2.  find all the children of this root resource and
+         *          try to match their path to the query
+         */
+        $basePath = Glob::getBasePath($query);
+        $baseResource = $this->resolveResource($basePath);
+
+        if ($baseResource !== null) {
+            $children = $this->getChildrenRecursive($baseResource);
+
+            $i = 0;
+            $found = false;
+
+            while ($i < $children->count() && !($found && $singleResult)) {
+                /** @var Resource $child */
+                $child = $children->get($i);
+
+                $found = Glob::match($child->getRepositoryPath(), $query);
+
+                if ($found) {
+                    $resources->add($child);
+                }
+
+                ++$i;
+            }
+        }
+
+        return $resources;
+    }
+
+    /**
+     * Add a given resource to the repository.
+     *
+     * @param string             $path
+     * @param FilesystemResource $resource
+     */
+    private function addResource($path, FilesystemResource $resource)
+    {
+        // Don't modify resources attached to other repositories
+        if ($resource->isAttached()) {
+            $resource = clone $resource;
+        }
+
+        $resource->attachTo($this, $path);
+
+        $this->store->set($path, $resource->getFilesystemPath());
+    }
+
+    /**
+     * Recursively find all the children of a given resource.
+     *
+     * @param Resource $resource The resource.
+     *
+     * @return ArrayResourceCollection The children.
+     */
+    private function getChildrenRecursive(Resource $resource)
+    {
+        $resources = new ArrayResourceCollection();
+
+        foreach ($this->getChildren($resource) as $child) {
+            $resources->merge($this->getChildrenRecursive($child)->toArray());
```

Comment:
I think this method should be optimized. In case of FilesystemResources, you should immediately do a recursive iteration instead of subsequent calls to `getChildren()`.